sdks/python/apache_beam/runners/portability/fn_api_runner_test.py

```diff
@@ -1083,6 +1131,34 @@ def __reduce__(self):
 def _unpickle_element_counter(name):
   return _pickled_element_counters[name]
 
+class EventRecorder(object):
+  """Used to be registered as a callback in bundle finalization.
+
+  The reason why records are written into a tmp file is, the in-memory dataset
+  cannot keep callback records when passing into one DoFn.
+  """
+  def __init__(self, tmp_dir):
+    self.tmp_dir = tmp_dir
+    self.file_path = os.path.join(self.tmp_dir, uuid.uuid4().hex + '.txt')
+
+  def record(self, content):
+    tmp_file = open(self.file_path, 'a')
+    tmp_file.write(content + '\n')
+    tmp_file.close()
+
+  def events(self):
+    tmp_file = open(self.file_path, 'r')
+    content = tmp_file.read()
+    tmp_file.close()
+    self.delete_file()
+    return content
```

Comment:
Just a thought as to API design: it's nice to make EventRecorder encapsulate its implementation by providing a symmetric API. For example, if I do `record("a"); record("b"); record("c")` I would expect to get `["a", "b", "c"]` back when asking for what was recorded, rather than a single newline-delimited string.